src/main/frontend/modules/outliner/core.cljs

```diff
@@ -192,229 +192,250 @@
+    :else
+    (or (get-new-id block left)
+        (get-new-id block (nth blocks (dec idx))))))
+
+(defn- insert-blocks-aux
+  [blocks target-block {:keys [sibling? replace-empty-target? keep-uuid? move? outliner-op]}]
+  (let [block-uuids (map :block/uuid blocks)
+        ids (set (map :db/id blocks))
+        uuids (zipmap block-uuids
+                      (if keep-uuid?
+                        block-uuids
+                        (repeatedly random-uuid)))
+        uuids (if replace-empty-target?
+                (assoc uuids (:block/uuid (first blocks)) (:block/uuid target-block))
+                uuids)
+        id->new-uuid (->> (map (fn [block] (when-let [id (:db/id block)]
+                                             [id (get uuids (:block/uuid block))])) blocks)
+                          (into {}))
+        target-page (or (:db/id (:block/page target-block))
+                        ;; target block is a page itself
+                        (:db/id target-block))
+        get-new-id (fn [block lookup]
+                     (cond
+                       (or (map? lookup) (vector? lookup))
+                       (when-let [uuid (if (and (vector? lookup) (= (first lookup) :block/uuid))
+                                         (get uuids (last lookup))
+                                         (get id->new-uuid (:db/id lookup)))]
+                         [:block/uuid uuid])
+
+                       (integer? lookup)
+                       lookup
+
+                       :else
+                       (throw (js/Error. (str "[insert-blocks] illegal lookup: " lookup ", block: " block)))))
+        indent-outdent? (= outliner-op :indent-outdent-blocks)]
+    (map-indexed (fn [idx {:block/keys [parent left] :as block}]
+                   (when-let [uuid (get uuids (:block/uuid block))]
+                     (let [top-level? (= (:block/level block) 1)
+                           outdented-block? (and indent-outdent?
+                                                 top-level?
+                                                 (not= (:block/parent block) (:block/parent target-block)))
+                           prev-hop (if outdented-block? (find-outdented-block-prev-hop block blocks) nil)
+                           left-exists-in-blocks? (contains? ids (:db/id (:block/left block)))
+                           parent (compute-block-parent block parent target-block prev-hop top-level? sibling? get-new-id)
+                           left (compute-block-left blocks block left target-block prev-hop idx replace-empty-target? left-exists-in-blocks? get-new-id)]
+                       (cond->
+                         (merge block {:block/uuid uuid
+                                       :block/page target-page
+                                       :block/parent parent
+                                       :block/left left})
+                         (not move?)
+                         (dissoc :db/id)))))
+                 blocks)))
+
```

Comment:
i think we should add more :pre :post asserts for `insert-blocks`, `delete-blocks`, `move-blocks`, etc ...
and doc all options, e.g. `insert-blocks` has options: [sibling? keep-uuid? move? outliner-op replace-empty-target?]

Reply:
Would also find this useful

Reply:
Thanks! Asserts and function docs are updated. Check https://github.com/logseq/logseq/pull/4880/files#diff-e66ebb001cf43bbf96ea541a0432ddda3d42f568e827be0a17c7c6ad753a0b64R432-R442.